common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+            val promise = Promise[W]
+            val desiredEntry = Entry(transid, ReadInProgress, promise.future)
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) flatMap { actualEntry =>
+                // ... and see what we get back
+
+                actualEntry.state.get match {
+                    case Cached => {
+                        logger.debug(this, s"Cached read $key")
+                        makeNoteOfCacheHit(key)
+                        actualEntry.value
+                    }
+
+                    case ReadInProgress => {
+                        if (actualEntry.transid == transid) {
+                            logger.debug(this, "Read initiated");
+                            makeNoteOfCacheMiss(key)
+                            listenForReadDone(key, actualEntry, future, promise)
+                            actualEntry.value
+
+                        } else {
+                            logger.debug(this, "Coalesced read")
+                            makeNoteOfCacheHit(key)
+                            actualEntry.value
+                        }
+                    }
+
+                    case WriteInProgress | InvalidateInProgress => {
+                        logger.debug(this, "Reading around a write in progress")
+                        makeNoteOfCacheMiss(key)
+                        future
+                    }
+                }
+            }
+
+        } else {
+            // not caching
+            future
+        }
+    }
+
+    protected def cacheUpdate(doc: W, key: Any, future: Future[Winfo])(
+        implicit transid: TransactionId, logger: Logging, ec: ExecutionContext): Future[Winfo] = {
+
+        if (cacheEnabled) {
+            logger.info(this, s"caching $key")
+
+            val desiredEntry = Entry(transid, WriteInProgress, Future { doc })
+
+            //
+            // try inserting our desired entry...
+            //
+            cache(key)(desiredEntry) map { actualEntry =>
+                // ... and see what we get back
+
+                if (transid == actualEntry.transid) {
+                    //
+                    // then we won the race, and now own the entry
+                    //
+                    logger.debug(this, s"Write initiated $key")
+                    listenForWriteDone(key, actualEntry, future)
+
+                } else {
+                    //
+                    // then either we lost the race, or there is an operation in progress on this key
+                    //
+                    // note that we don't care what that other operation is doing; unlike for lookups,
+                    // we don't need to do anything special (e.g. coalescing, as in the case of lookups)
+                    //
+                    logger.debug(this, s"Write under ${actualEntry.state.get} $key")
+                }
+            }
+
+        } else {
+            // not caching
+        }
```

Comment:
to me, this helps to preserve the understandability of the code; it also maintains a parallel structure between the lookups and updates.

i realize that we're getting into style, but this is something i feel the rest of the code base sorely lacks. it is too terse, and with very few comments explaining the flow. terseness is fine (and helpful), to a point. but, taken to far, and you produce code that only the core team understands.